aws/alb.go

```diff
@@ -0,0 +1,219 @@
+	scheme          string
+	subnets         []string
+	certificateARN  string
+	securityGroupID string
+	clusterName     string
+	targetGroupARNs []string
+}
+
+func createLoadBalancer(alb elbv2iface.ELBV2API, spec *createLoadBalancerSpec) (*LoadBalancer, error) {
+	var name = normalizeLoadBalancerName(spec.certificateARN)
+	params := &elbv2.CreateLoadBalancerInput{
+		Name:    aws.String(name),
+		Subnets: aws.StringSlice(spec.subnets),
+		Scheme:  aws.String(spec.scheme),
+		SecurityGroups: []*string{
+			aws.String(spec.securityGroupID),
+		},
+		Tags: []*elbv2.Tag{
+			{
+				Key:   aws.String(kubernetesClusterTag),
+				Value: aws.String(spec.clusterName),
+			},
+			{
+				Key:   aws.String(kubernetesCreatorTag),
+				Value: aws.String("kube-ingress-aws-controller"),
+			},
+		},
+	}
+	resp, err := alb.CreateLoadBalancer(params)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.LoadBalancers) < 1 {
+		return nil, errors.New("request to create ALB succeeded but returned no items")
+	}
+
+	newLoadBalancer := resp.LoadBalancers[0]
+	loadBalancerARN := aws.StringValue(newLoadBalancer.LoadBalancerArn)
+	newListener, err := createListener(alb, loadBalancerARN, spec)
+	if err != nil {
+		// TODO: delete just created LB?
+		return nil, err
+	}
+
+	return &LoadBalancer{
+		arn:      loadBalancerARN,
+		name:     name,
+		dnsName:  aws.StringValue(newLoadBalancer.DNSName),
+		listener: newListener,
+	}, nil
+}
+
+func normalizeLoadBalancerName(name string) string {
+	fields := strings.Split(name, "/")
+	if len(fields) >= 2 {
+		name = strings.Replace(fields[1], "-", "", -1)
+	}
+	if len(name) > 32 {
+		name = name[:32]
+	}
+	return name
+}
+
+func createListener(alb elbv2iface.ELBV2API, loadBalancerARN string, spec *createLoadBalancerSpec) (*loadBalancerListener, error) {
+	actions := make([]*elbv2.Action, len(spec.targetGroupARNs))
+	for i, tg := range spec.targetGroupARNs {
+		actions[i] = &elbv2.Action{TargetGroupArn: aws.String(tg), Type: aws.String(elbv2.ActionTypeEnumForward)}
+	}
+	params := &elbv2.CreateListenerInput{
+		Certificates: []*elbv2.Certificate{
+			{
+				CertificateArn: aws.String(spec.certificateARN),
+			},
+		},
+		LoadBalancerArn: aws.String(loadBalancerARN),
+		Port:            aws.Int64(443),
+		Protocol:        aws.String(elbv2.ProtocolEnumHttps),
+		DefaultActions:  actions,
+	}
+
+	resp, err := alb.CreateListener(params)
+	if err != nil {
+		return nil, err
+	}
+	if len(resp.Listeners) < 1 {
+		return nil, errors.New("request to create Listener succeeded but returned no items")
+	}
+	l := resp.Listeners[0]
+	return &loadBalancerListener{
+		arn:            aws.StringValue(l.ListenerArn),
+		port:           aws.Int64Value(l.Port),
+		certificateARN: spec.certificateARN,
+	}, nil
+}
+
+func createDefaultTargetGroup(alb elbv2iface.ELBV2API, clusterName string, vpcID string) ([]string, error) {
+	params := &elbv2.CreateTargetGroupInput{
+		HealthCheckPath: aws.String("/healthz"),
+		Port:            aws.Int64(9990),
```

Comment:
As discussed offline, both the target group and the security group creation is going to be moved to the kubernetes-aws bootstrapping process.